crates/ark/src/lsp/main_loop.rs

```diff
@@ -397,17 +399,47 @@ impl GlobalState {
 /// * - `response`: The response wrapped in a `anyhow::Result`. Errors are logged.
 /// * - `into_lsp_response`: A constructor for the relevant `LspResponse` variant.
 fn respond<T>(
-    response_tx: TokioUnboundedSender<anyhow::Result<LspResponse>>,
-    response: anyhow::Result<T>,
+    response_tx: TokioUnboundedSender<RequestResponse>,
+    response: impl FnOnce() -> anyhow::Result<T>,
     into_lsp_response: impl FnOnce(T) -> LspResponse,
 ) -> anyhow::Result<()> {
+    let mut crashed = false;
+
+    let response = std::panic::catch_unwind(std::panic::AssertUnwindSafe(response))
```

Comment:
I'm not entirely certain why we can `AssertUnwindSafe` here, maybe a comment?

Reply:
I've not done a thorough analysis but the main source of unsafety is interior mutability which we a priori don't use in request handlers. We're not in a position to guarantee safety here though.